Add IPM console

finikorg opened the pull request:
This PR adds IPM console which might be tested with up_squared_adsp.
Following options need to be enabled: CONSOLE & IPM and disable LOG_PRINTK

After that you can "flash" the board with:
```
$ west flash zephyr/boards/xtensa/up_squared_adsp/tools/up_squared_adsp_flash.sh <KEY>
```
And in another terminal
```
$ sudo boards/xtensa/up_squared_adsp/tools/mbterm.py 
*** Booting Zephyr OS build zephyr-v2.3.0-2086-g83a5af82007c  ***
Hello World! up_squared_adsp
```
Can be used with `sanitycheck` also, for example from inside zephyrproject:
```
$ zephyr/scripts/sanitycheck -M -c -p up_squared_adsp --device-test \
--west-flash="zephyr/boards/xtensa/up_squared_adsp/tools/up_squared_adsp_flash.sh,otc_private_key.pem" \
--device-serial-pty="$HOME/set_ambient zephyr/boards/xtensa/up_squared_adsp/tools/mbterm.py" \
-T zephyr/samples/hello_world \
-x="CONFIG_IPM=y" -x="CONFIG_CONSOLE=y" -x="CONFIG_LOG_PRINTK=n" -v
```
where `$HOME/set_ambient` is described here: https://github.com/zephyrproject-rtos/zephyr/pull/22724#issuecomment-627308907

finikorg committed:
console: ipm_console: Add console over IPM

Add console over Inter Processor Mailboxes (IPM).
This is useful for AMP processors like ADSP found on up_squared board.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
CODEOWNERS: Add ipm_console codeowner.

Add @finikorg as codeowner for ipm_console.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
soc: intel_apl_adsp: Add defaults for IPM and IPM console

Add defaults for IPM and IPM_CONSOLE when IPM and CONSOLE are
enabled.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
boards: up_squared_adsp: Add IPM terminal tool

Add tool which can access ADSP console over IPM using Python device
library and polling Doorbell registers.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
soc: intel_apl_adsp: Tweak adsp initialization

Tweak ADSP initialization to catch early console output.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

dcpleung commented on drivers/console/Kconfig:
```diff
@@ -168,6 +168,52 @@ config IPM_CONSOLE_STACK_SIZE
+
+config IPM_CONSOLE_ON_DEV_NAME
+	string "IPM device name used by console"
+	default "IPM_0"
```
Could you add a default which takes the target device in DTS, similar to the UART console?

finikorg replied:
I am thinking can I use something like: efaa9e311224f680f2900d5f3675104d6150e510. It is the same way IDC is used.

dcpleung replied:
That should work.

finikorg replied:
Added getting target console device through DTS, it works, can be probably improved since I am not very fluent in Device tree yaml programming.

dcpleung commented on drivers/console/ipm_console.c:
```diff
@@ -0,0 +1,86 @@
+	static size_t len;
+	int ret;
+
+	if (c != '\n' && len < sizeof(buf)) {
+		buf[len++] = c;
+		return c;
+	}
```
Wouldn't this miss the last out character if buffer is full?

1. Fill buffer to full. Now `len == sizeof(buf)` before `return`.
2. `console_out(c)` is called again with non-newline character.
3. Since `len == sizeof(buf)`, the assignment is skipped, and `ipm_send()` is called. But the character is not stored back into the buffer, and is not being sent thru IPM either. So this new character is lost.

finikorg replied:
Thanks for good catch, fixed

tyrantlizard77 reviewed (changes requested):
I'm super confused.

Something wrong with the existing IPM console driver?
See drivers/console/ipm_console_sender.c, drivers/console/ipm_console_receiver.c

finikorg commented:
> I'm super confused.
> 
> Something wrong with the existing IPM console driver?
> See drivers/console/ipm_console_sender.c, drivers/console/ipm_console_receiver.c

@andrewboie I can modify this `ipm_console_sender.c` to make it usable as a console. At the moment it is a kind of library used in `tests/drivers/ipm/` fot tests only as names `sender` and `receiver` implies (It can be probably moved to tests)

So it was easier to write proper ipm_console then modify those console senders and receivers and also tests.

@andyross also does not like the method used to send character: https://github.com/zephyrproject-rtos/zephyr/pull/22724/commits/2419dac1060bb11757519f17e43da7a694a27967#r379030695

andrewboie commented:
> At the moment it is a kind of library used in `tests/drivers/ipm/` fot tests only as names `sender` and `receiver` implies (It can be probably moved to tests)

No, it's not a test library. We used it to forward console output on Arduino 101 from the ARC CPU to the Lakemont, it was enabled by the default for that board.

The name "sender" is the side sending the console messages (on Arduino 101, the ARC CPU)
The name "receiver" is the code running on the other CPU (on Arudiuno 101, the Lakemont) which receives the console data and prints it.

I don't care if you rewrite it but if you want to rewrite how IPM console works to make it line-buffered you need to remove all of this code or amend it. Not have two completely different implementations for the same thing in-tree, that's a hard NACK.

finikorg commented:
> > At the moment it is a kind of library used in `tests/drivers/ipm/` fot tests only as names `sender` and `receiver` implies (It can be probably moved to tests)
> 
> No, it's not a test library. We used it to forward console output on Arduino 101 from the ARC CPU to the Lakemont, it was enabled by the default for that board.

But now it is enabled only by test.

> The name "sender" is the side sending the console messages (on Arduino 101, the ARC CPU)
> The name "receiver" is the code running on the other CPU (on Arudiuno 101, the Lakemont) which receives the console data and prints it.
> 
> I don't care if you rewrite it but if you want to rewrite how IPM console works to make it line-buffered you need to remove all of this code or amend it. Not have two completely different implementations for the same thing in-tree, that's a hard NACK.

As i said @andyross gave strong NACK with this method, even wrote small essay ;).

So at the moment nobody uses this code, if everybody OK I will remove it. I am also OK reusing console sender, I still have support in my python terminal tool with getting console characters in ID field. I just need for that enabling console sender and removing or modifying test.

finikorg committed:
drivers: console: Add DTS defaults

Add default which takes the target device from DTS similar way to UART
console.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
dts: Add binding for IPM console

Add IPM console DTS binding.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
boards: up_squared_adsp: Add IPM console to device tree

Add IPM console to the board.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

nashif commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -58,9 +58,6 @@ config CMSIS_V2_THREAD_MAX_STACK_SIZE
 config CMSIS_V2_THREAD_DYNAMIC_STACK_SIZE
```
why is this being removed? I know it was used in A101, but there is nothing here that is related to 101 and we have a test. So I would leave it, we might get users for this in the future.

finikorg replied:
@nashif Do you mean drop the commit 58e0c009ab5fb0df07d33fa365f655d1d0956b6f ?

nashif replied:
yes

finikorg committed:
sanitycheck: Allow to pass command with argument to ser_pty

By default, subprocess.Popen commands are supplied as a list of strings.
Using split() allows to use command with arguments, for example it is
possible to use following:

sanitycheck ... \
--device-serial-pty="set_ambient read_terminal.py" \
...

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
drivers: ipm-console: Remove unused config option

Remove now unused configuration option.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>
